(DOCSP-26772) [New Page] Manage User Sessions (#358)
mongodb/docs-app-services: source/users/sessions.txt

.. _manage-user-sessions:

====================
Manage User Sessions
====================

.. contents:: On this page
   :depth: 2

Overview
--------

App Services manages :ref:`user sessions <user-sessions>` with access
tokens and refresh tokens. An access token allows anyone that has the
token to use the session it's associated with for up to
{+user-session-expiration-time+}. A refresh token lets you generate a
new access token for a session even if the previous token has already
expired.

> 

Reviewer: i think this page could use a brief discussion of why you might want to get a user access token, in other words: what can you access with it? 

off top of my head, this would be:

- data api
- custom https endpoints
- graphql api

not sure if missing anything here.

.. note::

   The Realm SDKs automatically manage user access and refresh tokens
   for you when they connect to App Services.

.. _get-access-token:

Get a User Session Access Token
-------------------------------

You can authenticate a user and receive a user access token directly
over HTTPS or through a session managed by a Realm SDK. An access token

> expires

Reviewer: merp.

think some missing content here

Author: Lil bit extra - deleted

----------------------------------

By default, refresh tokens expire 60 days after they are issued. You can
configure this time for your App's refresh tokens to be anywhere between
30 minutes and 180 days. There are two exceptions:

- Anonymous user refresh tokens do not expire, though anonymous user
  accounts are deleted after 90 days of inactivity.

- Custom JWT refresh token expiration is determined by the ``exp`` value
  of the user's JWT.

.. note::

   You can currently only configure refresh token expiration with the
   Admin API.

Call your App's :admin-api-endpoint:`Set User Refresh Token Expiration
Time <operation/adminSetRefreshTokenExpiration>` Admin API endpoint with
the ``expiration_time_seconds`` field of the request body set to your
desired expiration time.

.. code-block:: bash

   curl -X POST \
     https://realm.mongodb.com/api/admin/v3.0/groups/{groupId}/apps/{appId}/security/refresh_token_expiration \
     --data-raw '{
       "expiration_time_seconds": 864000
     }'

If you successfully update the refresh token expiration time, the
endpoint returns a ``204`` response.

Verify & Decode a User Access Token
-----------------------------------

> You can decode an access token to verify that it is well-formed and has

Reviewer: ```suggestion
You can decode an access token to verify that it is correctly formed and has
```

word choice

a valid signature. The decoded response includes other information like
when the access token expires.

Call your App's :admin-api-endpoint:`Verify & decode an access token
<operation/adminVerifyUserAccessToken>` Admin API endpoint with the

> token in the ``token`` field of the request body.

Reviewer: correct me if my understanding of JWTs is correct here. 

while this approach is valid, you could also use any other JWT decoder to do the same thing. for example, i copied the example JWT below into https://jwt.io/ and got the output which you can find [at this link](https://jwt.io/#debugger-io?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJiYWFzX2RldmljZV9pZCI6IjAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMCIsImJhYXNfZG9tYWluX2lkIjoiNWNkYjEyNDA4ZTIzMmFjNGY5NTg3ZmU4IiwiZXhwIjoxNjc2NTExMjgyLCJpYXQiOjE2NzY1MDk0ODIsImlzcyI6IjYzZWQ4MTJhNDNiZTcyYzE3NmFhNWQyMyIsInN0aXRjaF9kZXZJZCI6IjAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMCIsInN0aXRjaF9kb21haW5JZCI6IjVjZGIxMjQwOGUyMzJhYzRmOTU4N2ZlOCIsInN1YiI6IjYzZWQ4MTJhNDNiZTcyYzE3NmFhNWQyMSIsInR5cCI6ImFjY2VzcyJ9.7kHO9wjWvIaD3VewDyPhLyb-oRc7wTYZdD9-hroF-H4).

with that in mind (if my understanding is correct), i think it'd make sense to note that you can use other JWT decoder tools/libraries to achieve the same effect.

Reviewer: also can you mention the HTTP status of a successful response here b/c you mention it for unsuccessful (which is 200 🤔)

Author: > you could also use any other JWT decoder to do the same thing

Sure you can decode the JWT but you can't validate the signature unless you have the 256 bit secret key. jwt.io tells me that it _might be_ a valid session token but can't confirm that it's authentic.

> can you mention the HTTP status of a successful response

Yeah good catch

        }'

   .. output::
      :language: json

      {
      	"sub": "63ed812a43be72c176aa5d21",
      	"exp": 1676511282,
      	"iat": 1676509482,
      	"iss": "63ed812a43be72c176aa5d23",
      	"domain_id": "5cdb12408e232ac4f9587fe8",
      	"device_id": "000000000000000000000000"
      }

If the token is invalid or expired, you'll receive a ``200`` response
with an error message:

.. code-block:: json

   {
     "error": "signature is invalid"
   }

End a User Session
------------------

Once a session has been established you cannot end it individually,
though you can :ref:`revoke all of a user's active sessions
<revoke-user-sessions>`.

You can effectively end a user session by deleting all copies of the
access token and refresh token. This prevents further access to the
session and requires the user to authenticate and start a new session

> to continue.

Reviewer: could you clarify where'd this be, in client-side code?

Author: It really depends - you could copy people's access tokens into MongoDB or automatically tweet them if you wanted to

> The ``User.logOut()`` method in the Realm SDKs deletes local copies of
> of the access token and refresh token and invalidates the refresh token
> so that it can't be used to get a new access token.

Reviewer: ```suggestion
Each Realm SDK's ``User.logOut()`` method deletes local copies of
of the access token and refresh token and invalidates the refresh token
so that it can't be used to get a new access token.
```

nit for clarity

Reviewer: also worth mentioning that each SDK's `App.deleteUser()` method would do same?

.. important::

>    Deleting and invalidating tokens does **not** end the active session

Reviewer: > Deleting and invalidating tokens 

where? in client side code?

Author: Wherever they are - it's all dependent on what you do with them

   on the server. If a malicious user copied the access token before it
   was deleted, they could use the token to make requests for up to
   {+user-session-expiration-time+} until it expires. These requests
   would appear to come from the user who "logged out".

.. _revoke-user-sessions:

Revoke a User's Sessions

> ------------------------

Reviewer: also deleting the user would do this. imo worth adding a sentence w ref to https://docs-atlas-staging.mongodb.com/atlas-app-services/docsworker-xlarge/get-access-token/users/manage/#delete-a-user

30 minutes and 180 days. There are two exceptions:

- Anonymous user refresh tokens do not expire, though anonymous user
  accounts are deleted after 90 days of inactivity.

- Custom JWT refresh token expiration is determined by the ``exp`` value
  of the user's JWT.

.. note::

   You can currently only configure refresh token expiration with the
   Admin API.

Call your App's :admin-api-endpoint:`Set User Refresh Token Expiration
Time <operation/adminSetRefreshTokenExpiration>` Admin API endpoint with
the ``expiration_time_seconds`` field of the request body set to your
desired expiration time.

.. code-block:: bash

   curl -X POST \
     https://realm.mongodb.com/api/admin/v3.0/groups/{groupId}/apps/{appId}/security/refresh_token_expiration \
     --data-raw '{
       "expiration_time_seconds": 864000
     }'

If you successfully update the refresh token expiration time, the
endpoint returns a ``204`` response.

Verify & Decode a User Access Token
-----------------------------------

You can decode an access token to verify that it has the correct format
and has a valid signature. The decoded response includes other
information like when the access token expires.

> Call your App's :admin-api-endpoint:`Verify & decode an access token

Reviewer: ```suggestion
Call your App's :admin-api-endpoint:`Verify and decode an access token
```

nit

      	"domain_id": "5cdb12408e232ac4f9587fe8",
      	"device_id": "000000000000000000000000"
      }

If the token is valid, you'll receive a ``200`` response.

If the token is valid but expired, the response indicates that the token
is expired rather than containing the decoded JWT.

.. code-block:: json

   "token expired"

If the token is invalid, you'll receive a ``401`` response that contains
an error message.

.. code-block:: json

   {
     "error": "signature is invalid"
   }

End a User Session
------------------

Once a session has been established you cannot end it individually,
though you can :ref:`revoke all of a user's active sessions
<revoke-user-sessions>`.

You can effectively end a user session by deleting all copies of the
access token and refresh token. This prevents further access to the
session and requires the user to authenticate and start a new session
to continue.

Each Realm SDK has a ``User.logOut()`` method that deletes local copies
of the access token and refresh token and invalidates the refresh token
so that it can't be used to get a new access token.

If a user account is deleted, either by an administrator or by the user,

> then all of the user's sessions are automatically revoked.

Reviewer: in this paragraph, consider ref to https://docs-atlas-staging.mongodb.com/atlas-app-services/docsworker-xlarge/get-access-token/users/manage/#delete-a-user
